Deriving pyclass/pymethods seems to break when inside of a macro-by-example.

### Bug Description

When placed inside of a macro code that compiles outside of the macro no longer compiles. The bug seems to be specifically around returning an instance of another pyclass from a method on another.

### Steps to Reproduce

```bash
maturin new repro --bindings pyo3
cd repro
cargo add sha3
cat >src/lib.rs <<'EOF'
∙ use pyo3::prelude::*;

macro_rules! impl_repro {
    ($name_a:ident, $name_b:ident) => {
        #[pyclass]
        struct $name_a {}

        #[pymethods]
        impl $name_a {
            fn finalize(&mut self) -> $name_b {
                $name_b {}
            }
        }

        #[pyclass]
        struct $name_b {}
    };
}

impl_repro!(MyClass1, MyClass2);
EOF
maturin develop --release
```

### Backtrace

```shell
error[E0425]: cannot find value `py` in this scope
  --> src/lib.rs:20:23
   |
20 | impl_repro!(MyClass1, MyClass2);
   |                       ^^^^^^^^ not found in this scope

error[E0425]: cannot find value `ret` in this scope
  --> src/lib.rs:20:23
   |
20 | impl_repro!(MyClass1, MyClass2);
   |                       ^^^^^^^^ not found in this scope
```
```


### Your operating system and version

Arch Linux (kernel version 6.9.9)

### Your Python version (`python --version`)

Python 3.12.4

### Your Rust version (`rustc --version`)

 rustc 1.80.0 (051478957 2024-07-21)

### Your PyO3 version

0.20.0

### How did you install python? Did you use a virtualenv?

python is installed via pacman
maturin is installed in a venv:
python -m venv .env
source .env/bin/activate
pip install maturin

### Additional Info

The original context for this is effectively attempting to get around the no-generics for pyclasses by templating out an implementation for a set of XOF functions where each hasher has a "Shaker" which absorbs data in and a "Sponge" which produces data, and you must finalize a Shaker to get a Sponge so I want to have finalize reset the state of the Shaker and return a Sponge object.

--- a/newsfragments/4382.fixed.md
+++ b/newsfragments/4382.fixed.md
@@ -1,0 +1,1 @@
+Fixed hygiene/span issues of emitted code which affected expansion in `macro_rules` context.

--- a/pyo3-macros-backend/src/method.rs
+++ b/pyo3-macros-backend/src/method.rs
@@ -720,9 +720,10 @@ impl<'a> FnSpec<'a> {
 
             // We must assign the output_span to the return value of the call,
             // but *not* of the call itself otherwise the spans get really weird
-            let ret_expr = quote! { let ret = #call; };
-            let ret_var = quote_spanned! {*output_span=> ret };
-            let return_conversion = quotes::map_result_into_ptr(quotes::ok_wrap(ret_var, ctx), ctx);
+            let ret_ident = Ident::new("ret", *output_span);
+            let ret_expr = quote! { let #ret_ident = #call; };
+            let return_conversion =
+                quotes::map_result_into_ptr(quotes::ok_wrap(ret_ident.to_token_stream(), ctx), ctx);
             quote! {
                 {
                     #ret_expr

--- a/pyo3-macros-backend/src/pyclass.rs
+++ b/pyo3-macros-backend/src/pyclass.rs
@@ -2081,7 +2081,7 @@ impl<'a> PyClassImplsBuilder<'a> {
         if attr.options.extends.is_none() {
             quote! {
                 impl #pyo3_path::IntoPy<#pyo3_path::PyObject> for #cls {
-                    fn into_py(self, py: #pyo3_path::Python) -> #pyo3_path::PyObject {
+                    fn into_py(self, py: #pyo3_path::Python<'_>) -> #pyo3_path::PyObject {
                         #pyo3_path::IntoPy::into_py(#pyo3_path::Py::new(py, self).unwrap(), py)
                     }
                 }

--- a/pyo3-macros-backend/src/quotes.rs
+++ b/pyo3-macros-backend/src/quotes.rs
@@ -27,5 +27,6 @@ pub(crate) fn map_result_into_ptr(result: TokenStream, ctx: &Ctx) -> TokenStream
         output_span,
     } = ctx;
     let pyo3_path = pyo3_path.to_tokens_spanned(*output_span);
-    quote_spanned! {*output_span=> #pyo3_path::impl_::wrap::map_result_into_ptr(py, #result) }
+    let py = syn::Ident::new("py", proc_macro2::Span::call_site());
+    quote_spanned! {*output_span=> #pyo3_path::impl_::wrap::map_result_into_ptr(#py, #result) }
 }

--- a/src/tests/hygiene/misc.rs
+++ b/src/tests/hygiene/misc.rs
@@ -37,3 +37,22 @@ fn append_to_inittab() {
 
     crate::append_to_inittab!(module_for_inittab);
 }
+
+macro_rules! macro_rules_hygiene {
+    ($name_a:ident, $name_b:ident) => {
+        #[crate::pyclass(crate = "crate")]
+        struct $name_a {}
+
+        #[crate::pymethods(crate = "crate")]
+        impl $name_a {
+            fn finalize(&mut self) -> $name_b {
+                $name_b {}
+            }
+        }
+
+        #[crate::pyclass(crate = "crate")]
+        struct $name_b {}
+    };
+}
+
+macro_rules_hygiene!(MyClass1, MyClass2);

